
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Lesson 5: Extracting Genes and Making Alignments &#8212; Gold Lab Training Manual</title>
    
  <link href="_static/css/theme.css" rel="stylesheet" />
  <link href="_static/css/index.c5995385ac14fb8791e8eb36b4908be2.css" rel="stylesheet" />

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/sphinx-book-theme.e8f53015daec13862f6db5e763c41738.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="_static/js/index.1c5a1a01449ed65a7b51.js">

    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/togglebutton.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Lesson 4: Finding Homologous Genes with BLAST" href="Lesson4.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
      <img src="_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Gold Lab Training Manual</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="intro.html">
   Welcome
  </a>
 </li>
</ul>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="Lesson0.html">
   Lesson 0: Configuring your computer
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="Lesson1.html">
   Lesson 1: Basic Command Line Skills
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="Lesson2.html">
   Lesson 2: Manipulating Text with Grep and Sed
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="Lesson3.html">
   Lesson 3: Git and GitHub
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="Lesson4.html">
   Lesson 4: Finding Homologous Genes with BLAST
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Lesson 5: Extracting Genes and Making Alignments
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="_sources/Lesson5.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/executablebooks/jupyter-book"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2FLesson5.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#where-we-left-off-with-lesson-4">
   5.0. Where we left off with Lesson 4
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introducing-samtools">
   5.1. Introducing Samtools
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#retrieving-genetic-data-with-samtools">
   5.2. Retrieving genetic data with Samtools
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#extracting-multiple-sequences-with-samtools">
     5.2.1. Extracting multiple sequences with Samtools
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#extracting-conserved-regions-from-multiple-genes-with-samtools">
   5.3. Extracting conserved regions from multiple genes with Samtools
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#homologous-genes-orthologs-and-paralogs">
   5.4. Homologous genes: Orthologs and Paralogs
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introducing-multiple-sequence-alignments">
   5.5 Introducing Multiple Sequence Alignments
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#programs-for-performing-msas">
     5.5.1. Programs for performing MSAs
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#making-a-msa-with-muscle">
   5.6. Making a MSA with MUSCLE
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#extract-sequence-ids-with-awk-gsed-and-samtools">
     5.6.1. Extract sequence IDs with awk, gsed, and Samtools
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#align-sequences-with-muscle">
     5.6.2. Align sequences with MUSCLE
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#visualize-your-msa">
   5.7. Visualize your MSA
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#cleanup-option-1-only-use-the-aligned-regions-of-the-blast-hits">
   5.8 Cleanup option 1: Only use the aligned regions of the BLAST hits
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#cleanup-option-2-identify-and-isolate-conserved-domains">
   5.9 Cleanup option 2: Identify and Isolate Conserved Domains
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#introducing-conserved-domains">
     5.9.1. Introducing conserved domains
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#introducing-the-pfam-database">
     5.9.2. Introducing the PFAM database
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#find-conserved-domains-with-the-pfam-web-server">
     5.9.3. Find conserved domains with the PFAM web server
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#upload-changes-to-your-github-repository">
   5.10. Upload changes to your GitHub repository
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <!-- #region -->
<div class="section" id="lesson-5-extracting-genes-and-making-alignments">
<h1>Lesson 5: Extracting Genes and Making Alignments<a class="headerlink" href="#lesson-5-extracting-genes-and-making-alignments" title="Permalink to this headline">¶</a></h1>
<p>© David Gold. Except where the source is noted, this work is licensed under a <a class="reference external" href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution License CC-BY 4.0</a>.</p>
<p>Let’s start by checking for any changes in the course material on GitHub:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>git fetch upstream
git checkout master
git merge upstream/master
</pre></div>
</div>
<div class="section" id="where-we-left-off-with-lesson-4">
<h2>5.0. Where we left off with Lesson 4<a class="headerlink" href="#where-we-left-off-with-lesson-4" title="Permalink to this headline">¶</a></h2>
<p>In the last lesson, we used <code class="docutils literal notranslate"><span class="pre">blastp</span></code> to compare several query sequences against the brewer’s yeast (<em>Saccharomyces cerevisiae</em>) proteome. Let’s take a look at those results:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="o">~/</span><span class="n">git</span><span class="o">/</span><span class="n">Gold_Lab_Training</span><span class="o">/</span><span class="n">Additional_Materials</span>
<span class="n">head</span> <span class="n">Lesson_4_Results</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>Terminal should return the following:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>qseqid	sseqid	pident	length	mismatch	gapopen	qstart	qend	sstart	send	evalue	bitscore
Ectocarpus_siliculosus_CBN76684.1_Sterol_methyltransferase	sp|P25087|ERG6_YEAST_Sterol_24-C-methyltransferase	43.296	358	182	3	75	413	13	368	2.78e-100	301
</pre></div>
</div>
<p>Here are the results in a table view for easier interpretation:</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="text-align:left head"><p>qseqid</p></th>
<th class="text-align:left head"><p>sseqid</p></th>
<th class="text-align:left head"><p>pident</p></th>
<th class="text-align:left head"><p>length</p></th>
<th class="text-align:left head"><p>mismatch</p></th>
<th class="text-align:left head"><p>gapopen</p></th>
<th class="text-align:left head"><p>qstart</p></th>
<th class="text-align:left head"><p>qend</p></th>
<th class="text-align:left head"><p>sstart</p></th>
<th class="text-align:left head"><p>send</p></th>
<th class="text-align:left head"><p>evalue</p></th>
<th class="text-align:left head"><p>bitscore</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-align:left"><p>Ectocarpus_siliculosus_CBN76684.1_Sterol_methyltransferase</p></td>
<td class="text-align:left"><p>sp|P25087|ERG6_YEAST_Sterol_24-C-methyltransferase</p></td>
<td class="text-align:left"><p>43.296</p></td>
<td class="text-align:left"><p>358</p></td>
<td class="text-align:left"><p>182</p></td>
<td class="text-align:left"><p>3</p></td>
<td class="text-align:left"><p>75</p></td>
<td class="text-align:left"><p>413</p></td>
<td class="text-align:left"><p>13</p></td>
<td class="text-align:left"><p>368</p></td>
<td class="text-align:left"><p>2.78e-100</p></td>
<td class="text-align:left"><p>301</p></td>
</tr>
</tbody>
</table>
<p>This is what the header IDs mean:</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="text-align:center head"><p>Column header</p></th>
<th class="text-align:center head"><p>Meaning</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-align:center"><p>qseqid</p></td>
<td class="text-align:center"><p>query (e.g., gene) sequence id</p></td>
</tr>
<tr class="row-odd"><td class="text-align:center"><p>sseqid</p></td>
<td class="text-align:center"><p>subject (e.g., reference genome) sequence id</p></td>
</tr>
<tr class="row-even"><td class="text-align:center"><p>pident</p></td>
<td class="text-align:center"><p>percentage of identical matches</p></td>
</tr>
<tr class="row-odd"><td class="text-align:center"><p>length</p></td>
<td class="text-align:center"><p>alignment length</p></td>
</tr>
<tr class="row-even"><td class="text-align:center"><p>mismatch</p></td>
<td class="text-align:center"><p>number of mismatches</p></td>
</tr>
<tr class="row-odd"><td class="text-align:center"><p>gapopen</p></td>
<td class="text-align:center"><p>number of gap openings</p></td>
</tr>
<tr class="row-even"><td class="text-align:center"><p>qstart</p></td>
<td class="text-align:center"><p>start of alignment in query</p></td>
</tr>
<tr class="row-odd"><td class="text-align:center"><p>qend</p></td>
<td class="text-align:center"><p>end of alignment in query</p></td>
</tr>
<tr class="row-even"><td class="text-align:center"><p>sstart</p></td>
<td class="text-align:center"><p>start of alignment in subject</p></td>
</tr>
<tr class="row-odd"><td class="text-align:center"><p>send</p></td>
<td class="text-align:center"><p>end of alignment in subject</p></td>
</tr>
<tr class="row-even"><td class="text-align:center"><p>evalue</p></td>
<td class="text-align:center"><p>expect value</p></td>
</tr>
<tr class="row-odd"><td class="text-align:center"><p>bitscore</p></td>
<td class="text-align:center"><p>bit score</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="introducing-samtools">
<h2>5.1. Introducing Samtools<a class="headerlink" href="#introducing-samtools" title="Permalink to this headline">¶</a></h2>
<p>As a reminder, one of our query sequences (Ectocarpus_siliculosus_CBN76684.1_Sterol_methyltransferase) had a single match in the yeast database (sp|P25087|ERG6_YEAST_Sterol_24-C-methyltransferase). If we want to see what this match looks like, we could open the relevant fasta file in BBEdit and search for it, but that won’t work with very large fasta files. Instead we will use a program called <a class="reference external" href="http://www.htslib.org/"><strong>Samtools</strong></a>. Samtools is primarily used for manipulating SAM and BAM files, a common format for high-throughput sequencing data. But Samtools has some useful functions for working with fasta files more broadly.</p>
<p>We can install Samtools with Homebrew:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">brew</span> <span class="n">install</span> <span class="n">samtools</span>
</pre></div>
</div>
</div>
<div class="section" id="retrieving-genetic-data-with-samtools">
<h2>5.2. Retrieving genetic data with Samtools<a class="headerlink" href="#retrieving-genetic-data-with-samtools" title="Permalink to this headline">¶</a></h2>
<p>Before we can extract data from a fasta file using Samtools, we need to <strong>index</strong> the file. We can do that with Samtools’ <code class="docutils literal notranslate"><span class="pre">faidx</span></code> command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">samtools</span> <span class="n">faidx</span> <span class="n">Lesson_4_Yeast_Proteome</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
<p>There is now an index file (Lesson_4_Yeast_Proteome.fasta.fai) added to your folder. Indexing allows Samtools to easily navigate fasta files of any size.</p>
<p>To extract the gene we found with BLAST, we also use the <code class="docutils literal notranslate"><span class="pre">faidx</span></code> command. But this time we add the name of the sequence we’re interested in. <strong>Make sure to wrap the name in quotes so Terminal does not try to interpret characters in the sequence name as regular expressions</strong>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">samtools</span> <span class="n">faidx</span> <span class="n">Lesson_4_Yeast_Proteome</span><span class="o">.</span><span class="n">fasta</span> <span class="s2">&quot;sp|P25087|ERG6_YEAST_Sterol_24-C-methyltransferase&quot;</span>
</pre></div>
</div>
<p>Terminal will return the relevant amino acid sequence:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;sp|P25087|ERG6_YEAST_Sterol_24-C-methyltransferase
MSETELRKRQAQFTRELHGDDIGKKTGLSALMSKNNSAQKEAVQKYLRNWDGRTDKDAEE
RRLEDYNEATHSYYNVVTDFYEYGWGSSFHFSRFYKGESFAASIARHEHYLAYKAGIQRG
DLVLDVGCGVGGPAREIARFTGCNVIGLNNNDYQIAKAKYYAKKYNLSDQMDFVKGDFMK
MDFEENTFDKVYAIEATCHAPKLEGVYSEIYKVLKPGGTFAVYEWVMTDKYDENNPEHRK
IAYEIELGDGIPKMFHVDVARKALKNCGFEVLVSEDLADNDDEIPWYYPLTGEWKYVQNL
ANLATFFRTSYLGRQFTTAMVTVMEKLGLAPEGSKEVTAALENAAVGLVAGGKSKLFTPM
MLFVARKPENAETPSQTSQEATQ
</pre></div>
</div>
<p>If you want to save this sequence you can redirect it to an output file (we’ll call it ERG6.fasta) with the <code class="docutils literal notranslate"><span class="pre">&gt;</span></code> command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">samtools</span> <span class="n">faidx</span> <span class="n">Lesson_4_Yeast_Proteome</span><span class="o">.</span><span class="n">fasta</span> <span class="s2">&quot;sp|P25087|ERG6_YEAST_Sterol_24-C-methyltransferase&quot;</span> <span class="o">&gt;</span> <span class="n">Lesson_5_1_ERG6</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
<p>Perhaps you only want to see the part of the yeast protein that matched our query. Using the <code class="docutils literal notranslate"><span class="pre">sstart</span></code> (start of alignment in subject) and <code class="docutils literal notranslate"><span class="pre">send</span></code> (end of alignment in subject) columns, you’ll see that the match encompasses amino acids #13-368.</p>
<p>We can extract these particular sequences using the previous <code class="docutils literal notranslate"><span class="pre">samtools</span> <span class="pre">faidx</span></code> command, but adding a colon (<code class="docutils literal notranslate"><span class="pre">:</span></code>) followed by the region of interest:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">samtools</span> <span class="n">faidx</span> <span class="n">Lesson_4_Yeast_Proteome</span><span class="o">.</span><span class="n">fasta</span> <span class="s2">&quot;sp|P25087|ERG6_YEAST_Sterol_24-C-methyltransferase&quot;</span><span class="p">:</span><span class="mi">13</span><span class="o">-</span><span class="mi">368</span>
</pre></div>
</div>
<p>Terminal will return the following:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;sp|P25087|ERG6_YEAST_Sterol_24-C-methyltransferase:13-368
FTRELHGDDIGKKTGLSALMSKNNSAQKEAVQKYLRNWDGRTDKDAEERRLEDYNEATHS
YYNVVTDFYEYGWGSSFHFSRFYKGESFAASIARHEHYLAYKAGIQRGDLVLDVGCGVGG
PAREIARFTGCNVIGLNNNDYQIAKAKYYAKKYNLSDQMDFVKGDFMKMDFEENTFDKVY
AIEATCHAPKLEGVYSEIYKVLKPGGTFAVYEWVMTDKYDENNPEHRKIAYEIELGDGIP
KMFHVDVARKALKNCGFEVLVSEDLADNDDEIPWYYPLTGEWKYVQNLANLATFFRTSYL
GRQFTTAMVTVMEKLGLAPEGSKEVTAALENAAVGLVAGGKSKLFTPMMLFVARKP
</pre></div>
</div>
<p>Compare it to the previous result. Notice how amino acids have been trimmed from the beginning and end of the sequence.</p>
<div class="section" id="extracting-multiple-sequences-with-samtools">
<h3>5.2.1. Extracting multiple sequences with Samtools<a class="headerlink" href="#extracting-multiple-sequences-with-samtools" title="Permalink to this headline">¶</a></h3>
<p>Let’s say you have a list of sequences you would like to retrieve from your fasta file. There are two ways to do that. The first is to append a list of sequence IDs in the <code class="docutils literal notranslate"><span class="pre">samtools</span> <span class="pre">faidx</span></code> command. For example, let’s say you wanted to extract the following proteins from the yeast proteome:</p>
<ul class="simple">
<li><p>sp|Q08054|CSI2_YEAST_Chitin_synthase_3_complex_protein_CSI2</p></li>
<li><p>sp|P24813|AP2_YEAST_AP-1-like_transcription_factor_YAP2</p></li>
<li><p>sp|Q01389|BCK1_YEAST_Serine/threonine-protein_kinase_BCK1/SLK1/SSP31</p></li>
</ul>
<p>You could use the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">samtools</span> <span class="n">faidx</span> <span class="n">Lesson_4_Yeast_Proteome</span><span class="o">.</span><span class="n">fasta</span> \
<span class="s2">&quot;sp|Q08054|CSI2_YEAST_Chitin_synthase_3_complex_protein_CSI2&quot;</span> \
<span class="s2">&quot;sp|P24813|AP2_YEAST_AP-1-like_transcription_factor_YAP2&quot;</span> \
<span class="s2">&quot;sp|Q01389|BCK1_YEAST_Serine/threonine-protein_kinase_BCK1/SLK1/SSP31&quot;</span>
</pre></div>
</div>
<p>You could also extract portions of each sequence using the technique described in part 5.2:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">samtools</span> <span class="n">faidx</span> <span class="n">Lesson_4_Yeast_Proteome</span><span class="o">.</span><span class="n">fasta</span> \
<span class="s2">&quot;sp|Q08054|CSI2_YEAST_Chitin_synthase_3_complex_protein_CSI2&quot;</span><span class="p">:</span><span class="mi">2</span><span class="o">-</span><span class="mi">11</span> \
<span class="s2">&quot;sp|P24813|AP2_YEAST_AP-1-like_transcription_factor_YAP2&quot;</span><span class="p">:</span><span class="mi">15</span><span class="o">-</span><span class="mi">50</span> \
<span class="s2">&quot;sp|Q01389|BCK1_YEAST_Serine/threonine-protein_kinase_BCK1/SLK1/SSP31&quot;</span><span class="p">:</span><span class="mi">25</span><span class="o">-</span><span class="mi">40</span>
</pre></div>
</div>
<p>This approach works if you have a few sequences, but it can become cumbersome if you are trying to extract large numbers of sequences. An alternative approach is to put your list of sequence IDs into a text file. Execute the code below to make a text file called ‘Desired_Sequences.txt’ :</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">echo</span> <span class="o">-</span><span class="n">e</span> <span class="s1">&#39;sp|Q08054|CSI2_YEAST_Chitin_synthase_3_complex_protein_CSI2</span><span class="se">\n</span><span class="s1">sp|P24813|AP2_YEAST_AP-1-like_transcription_factor_YAP2</span><span class="se">\n</span><span class="s1">sp|Q01389|BCK1_YEAST_Serine/threonine-protein_kinase_BCK1/SLK1/SSP31&#39;</span> <span class="o">&gt;</span> <span class="n">Desired_Sequences</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>There are three sequence IDs in ‘Desired_Sequences.txt’. You can see them with the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">head</span> <span class="n">Desired_Sequences</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>Before using this list in Samtools, we need to add quotes to each sequence ID. We can easily do that using the <code class="docutils literal notranslate"><span class="pre">sed</span></code> command. We will first replace the start of each line (grep symbol = <code class="docutils literal notranslate"><span class="pre">^</span></code>) with a <code class="docutils literal notranslate"><span class="pre">&quot;</span></code> character. We will then replace the end of each line (grep symbol = <code class="docutils literal notranslate"><span class="pre">$</span></code>) with a <code class="docutils literal notranslate"><span class="pre">&quot;</span></code> character:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">gsed</span> <span class="o">-</span><span class="n">i</span> <span class="s1">&#39;s/^/&quot;/g&#39;</span> <span class="n">Desired_Sequences</span><span class="o">.</span><span class="n">txt</span>
<span class="n">gsed</span> <span class="o">-</span><span class="n">i</span> <span class="s1">&#39;s/$/&quot;/g&#39;</span> <span class="n">Desired_Sequences</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>Now that your text file is ready, you can use it to extract sequences with samtools. To do this we will need the program <code class="docutils literal notranslate"><span class="pre">xargs</span></code>, which is included in Unix-based operating systems. The job of xargs is to execute commands based on standard input. We will use xargs to feed the list into samtools (using the <code class="docutils literal notranslate"><span class="pre">&lt;</span></code> symbol) and output the list into a new file (using the symbol <code class="docutils literal notranslate"><span class="pre">&gt;</span></code>):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">xargs</span> <span class="n">samtools</span> <span class="n">faidx</span> <span class="n">Lesson_4_Yeast_Proteome</span><span class="o">.</span><span class="n">fasta</span> <span class="o">&lt;</span> <span class="n">Desired_Sequences</span><span class="o">.</span><span class="n">txt</span> <span class="o">&gt;</span> <span class="n">Desired_Sequences</span><span class="o">.</span><span class="n">Output</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
<p>You can look at “Desired_Sequences.txt” and “Desired_Sequences.Output.fasta” in BBEdit to verify that your list of 3 sequences ended up in the output fasta file. You can easily access these files by opening your current directory using Finder:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">open</span> <span class="o">.</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="extracting-conserved-regions-from-multiple-genes-with-samtools">
<h2>5.3. Extracting conserved regions from multiple genes with Samtools<a class="headerlink" href="#extracting-conserved-regions-from-multiple-genes-with-samtools" title="Permalink to this headline">¶</a></h2>
<p>In the “Additional_Materials” folder you should find a file called “Lesson_5_Query_Domain.fasta”. It contains the a seripauperin protein from the yeast <em>Saccharomyces arboricola</em> (a different species than the one in our database). Scientists do not fully understand the purpose of seripauperin genes, although they appear to be expressed during alcohol fermentation.</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;PAU5_Saccharomyces_arboricola_EJS43917.1
MVKLTSIAAGVAAIAAGASAATTTLAQSDEKVNLVELGVYVSDIRAHMAQYYLFQAAHPTETYPIEVAEA
VFNYGDFTTMLTGIAADQVTRMITGVPWYSTRLRPAISSALSKDGIYTIAN
</pre></div>
</div>
<p>Let’s use BLAST to look for the homologous gene in our <em>Saccharomyces cerevisiae</em> proteome database:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">blastp</span> <span class="o">-</span><span class="n">query</span> <span class="n">Lesson_5_Query_Domain</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">db</span> <span class="n">Yeast</span> <span class="o">-</span><span class="n">outfmt</span> <span class="mi">6</span> <span class="o">-</span><span class="n">evalue</span> <span class="mf">10e-5</span> <span class="o">-</span><span class="n">out</span> <span class="n">Lesson_5_2_BLAST_Results</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>Let’s take a look at the results:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">head</span> <span class="o">-</span><span class="n">n</span> <span class="mi">100</span> <span class="n">Lesson_5_2_BLAST_Results</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>PAU5_Saccharomyces_arboricola_EJS43917.1	sp|Q08322|PAU20_YEAST_Seripauperin-20	90.083	121	11	1	1	121	1	120	5.92e-69	201
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P35994|PAU16_YEAST_Seripauperin-16	87.805	123	13	1	1	121	1	123	3.90e-68	199
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P0CE92|PAU8_YEAST_Seripauperin-8	90.909	121	10	1	1	121	1	120	3.10e-67	196
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P0CE93|PAU11_YEAST_Seripauperin-11	90.909	121	10	1	1	121	1	120	3.10e-67	196
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|Q3E770|PAU9_YEAST_Seripauperin-9	90.909	121	10	1	1	121	1	120	3.10e-67	196
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P0CE91|PAU18_YEAST_Seripauperin-18	90.909	121	10	1	1	121	1	120	4.08e-67	196
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P0CE90|PAU6_YEAST_Seripauperin-6	90.909	121	10	1	1	121	1	120	4.08e-67	196
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P0CE89|PAU14_YEAST_Seripauperin-14	90.083	121	11	1	1	121	1	120	6.33e-67	196
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P0CE88|PAU1_YEAST_Seripauperin-1	90.083	121	11	1	1	121	1	120	6.33e-67	196
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P32612|PAU2_YEAST_Seripauperin-2	91.736	121	9	1	1	121	1	120	1.05e-66	195
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P53427|PAU4_YEAST_Seripauperin-4	89.256	121	12	1	1	121	1	120	1.08e-66	195
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P38725|PAU13_YEAST_Seripauperin-13	90.083	121	11	1	1	121	1	120	1.34e-66	195
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P0CE85|PAU19_YEAST_Seripauperin-19	88.333	120	12	1	1	118	1	120	1.93e-66	194
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P53343|PAU12_YEAST_Seripauperin-12	90.083	121	11	1	1	121	1	120	2.97e-66	194
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|Q07987|PAU23_YEAST_Seripauperin-23	86.667	120	14	1	1	118	1	120	3.74e-66	194
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P0CE87|PAU22_YEAST_Seripauperin-22	88.333	120	12	1	1	118	41	160	3.87e-66	195
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P0CE86|PAU21_YEAST_Seripauperin-21	88.333	120	12	1	1	118	41	160	3.87e-66	195
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P40585|PAU15_YEAST_Seripauperin-15	87.500	120	13	1	1	118	1	120	5.13e-66	193
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|Q03050|PAU10_YEAST_Seripauperin-10	90.083	121	11	1	1	121	1	120	9.21e-66	192
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P25610|PAU3_YEAST_Seripauperin-3	89.167	120	11	1	1	118	1	120	1.06e-65	192
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P38155|PAU24_YEAST_Seripauperin-24	89.256	121	12	1	1	121	1	120	1.13e-65	192
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P43575|PAU5_YEAST_Seripauperin-5	89.344	122	12	1	1	121	1	122	2.14e-65	192
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|Q12370|PAU17_YEAST_Seripauperin-17	86.667	120	14	1	1	118	1	120	4.55e-65	191
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P47179|DAN4_YEAST_Cell_wall_protein_DAN4	78.889	90	19	0	29	118	30	119	4.20e-49	165
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P47178|DAN1_YEAST_Cell_wall_protein_DAN1	71.134	97	28	0	22	118	23	119	8.39e-46	148
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P39545|PAU7_YEAST_Seripauperin-7	89.091	55	5	1	1	54	1	55	1.14e-15	63.9
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P10863|TIR1_YEAST_Cold_shock-induced_protein_TIR1	27.174	92	56	2	27	109	19	108	2.05e-08	48.5
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P27654|TIP1_YEAST_Temperature_shock-inducible_protein_1	30.864	81	51	2	36	111	28	108	2.42e-08	48.1
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P40552|TIR3_YEAST_Cell_wall_protein_TIR3	30.769	91	58	2	31	116	26	116	3.71e-08	47.8
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|P33890|TIR2_YEAST_Cold_shock-induced_protein_TIR2	25.532	94	59	2	27	111	19	110	1.48e-06	43.5
PAU5_Saccharomyces_arboricola_EJS43917.1	sp|Q12218|TIR4_YEAST_Cell_wall_protein_TIR4	29.670	91	52	4	27	107	19	107	1.51e-05	40.8
</pre></div>
</div>
<p>This time there are many hits in the BLAST database! Some of these matches could be due to chance, but it is more likely that these genes are all related to each other. It turns out that homologous genes don’t just exist between species; homologous genes exist within species as well!</p>
</div>
<div class="section" id="homologous-genes-orthologs-and-paralogs">
<h2>5.4. Homologous genes: Orthologs and Paralogs<a class="headerlink" href="#homologous-genes-orthologs-and-paralogs" title="Permalink to this headline">¶</a></h2>
<p>Gene duplications, recombination, and gene loss events over the course of evolution result in different numbers of homologous genes in different organisms. This creates a pattern of <strong>orthologs</strong> (genes generated from speciation events) and <strong>paralogs</strong> (genes generated from duplication events). Since all of the genes in our database come from one organism (<em>Saccharomyces cerevisiae</em>), these genes are presumably all paralogs of each other.</p>
</div>
<div class="section" id="introducing-multiple-sequence-alignments">
<h2>5.5 Introducing Multiple Sequence Alignments<a class="headerlink" href="#introducing-multiple-sequence-alignments" title="Permalink to this headline">¶</a></h2>
<p>To figure out homologous genes are related to each other (e.g. figuring out which genes are orthologs and which are paralogs), we need to reconstruct their evolutionary history. The first step in making an evolutionary tree is <strong>aligning</strong> the homologous genes. When we align more than two genes to each other it is called a <strong>multiple sequence alignment (MSA)</strong>. Here is an example MSA from <a class="reference external" href="https://en.wikipedia.org/wiki/Multiple_sequence_alignment">Wikipedia</a>:</p>
<p><img alt="multiple sequence alignment" src="https://upload.wikimedia.org/wikipedia/commons/7/79/RPLP0_90_ClustalW_aln.gif" /></p>
<p>Each row in the MSA represents a sequence. Each column represents a (supposedly) conserved amino acid across the homologous genes. Highly conserved parts of the protein are colored based on the most common amino acid. <strong>Each column in a MSA represents a hypothesis</strong>, suggesting that the amino acids in that column all trace back to an ancestral amino acid from the ancestral protein. You can make a MSA from sequences that are not homologous, but the results would be meaningless. It is therefore important to start with sequences you think are homologous.</p>
<div class="section" id="programs-for-performing-msas">
<h3>5.5.1. Programs for performing MSAs<a class="headerlink" href="#programs-for-performing-msas" title="Permalink to this headline">¶</a></h3>
<p>There are many different programs that perform multiple sequence alignments. All of them try to find the best (i.e. true) alignment using different methodologies:</p>
<ul class="simple">
<li><p>Progressive alignments (<a class="reference external" href="http://www.clustal.org/">CLUSTAL</a>), <a class="reference external" href="https://mafft.cbrc.jp/alignment/software/">MAFFT</a>): start with most similar sequences and build out. These are fast but not guaranteed to converge on the best alignment.</p></li>
<li><p>Iterative methods (<a class="reference external" href="https://www.drive5.com/muscle/">MUSCLE</a>): repeatedly realign the initial sequences as well as adding new sequences to the growing alignment.</p></li>
<li><p>Phylogeny-aware methods (<a class="reference external" href="http://wasabiapp.org/software/pagan/phylogenetic_multiple_alignment/">PAGAN</a>; <a class="reference external" href="http://wasabiapp.org/software/prank/">Prank</a>): provide a starting tree.</p></li>
<li><p>Motif finding methods (<a class="reference external" href="http://meme-suite.org/">MEME</a>) search for short, highly conserved patterns within the larger alignment and build out from there.</p></li>
</ul>
</div>
</div>
<div class="section" id="making-a-msa-with-muscle">
<h2>5.6. Making a MSA with MUSCLE<a class="headerlink" href="#making-a-msa-with-muscle" title="Permalink to this headline">¶</a></h2>
<p>For this next part, we’re going to make an alignment using the <em>Saccharomyces cerevisiae</em> Seripauperin proteins. We will use the BLAST results from “Lesson_5_2_BLAST_Results.txt” to identify likely Seripauperin proteins.</p>
<div class="section" id="extract-sequence-ids-with-awk-gsed-and-samtools">
<h3>5.6.1. Extract sequence IDs with awk, gsed, and Samtools<a class="headerlink" href="#extract-sequence-ids-with-awk-gsed-and-samtools" title="Permalink to this headline">¶</a></h3>
<p>The sequence IDs of the Seripauperin proteins are in the second column of “Lesson_5_2_BLAST_Results.txt”. To extract the list of sequence IDs, we’re going to use <code class="docutils literal notranslate"><span class="pre">awk</span></code> along with <code class="docutils literal notranslate"><span class="pre">xargs</span></code>. Awk is a language used for manipulating text data. When we have a tab-delimited file, the <code class="docutils literal notranslate"><span class="pre">print</span></code> function in awk is a powerful way to pull out columns:</p>
<p>Extract the second column from “Lesson_5_2_BLAST_Results.txt” with awk:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">awk</span> <span class="s1">&#39;{print $2}&#39;</span> <span class="n">Lesson_5_2_BLAST_Results</span><span class="o">.</span><span class="n">txt</span> <span class="o">&gt;</span> <span class="n">Temp_BLAST_List</span>
</pre></div>
</div>
<p>Add quotes to the start and end of sequence IDs with <code class="docutils literal notranslate"><span class="pre">gsed</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">gsed</span> <span class="o">-</span><span class="n">i</span> <span class="s1">&#39;s/^/&quot;/g&#39;</span> <span class="n">Temp_BLAST_List</span>
<span class="n">gsed</span> <span class="o">-</span><span class="n">i</span> <span class="s1">&#39;s/$/&quot;/g&#39;</span> <span class="n">Temp_BLAST_List</span>
</pre></div>
</div>
<p>And finally use the list “Temp_BLAST_List” to extract the sequences with Samtools:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">xargs</span> <span class="n">samtools</span> <span class="n">faidx</span> <span class="n">Lesson_4_Yeast_Proteome</span><span class="o">.</span><span class="n">fasta</span> <span class="o">&lt;</span> <span class="n">Temp_BLAST_List</span> <span class="o">&gt;</span> <span class="n">Lesson_5_3_PAU5_BLAST_Hits</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
<p>You can quickly check that everything worked using <code class="docutils literal notranslate"><span class="pre">head</span></code>, and then delete the temporary list file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">head</span> <span class="n">Lesson_5_3_PAU5_BLAST_Hits</span><span class="o">.</span><span class="n">fasta</span>
<span class="n">rm</span> <span class="o">-</span><span class="n">i</span> <span class="n">Temp_BLAST_List</span>
</pre></div>
</div>
<p>You can see how many sequences are in this file by counting the number of greater-than (<code class="docutils literal notranslate"><span class="pre">&gt;</span></code>) symbols using <code class="docutils literal notranslate"><span class="pre">grep</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">grep</span> <span class="o">-</span><span class="n">c</span> <span class="s2">&quot;&gt;&quot;</span> <span class="n">Lesson_5_3_PAU5_BLAST_Hits</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
<p>The file “Lesson_5_3_PAU5_BLAST_Hits.fasta” contains a fasta file with 35 putative Seripauperin proteins!</p>
</div>
<div class="section" id="align-sequences-with-muscle">
<h3>5.6.2. Align sequences with MUSCLE<a class="headerlink" href="#align-sequences-with-muscle" title="Permalink to this headline">¶</a></h3>
<p>Before using MUSCLE, you have to install it:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">brew</span> <span class="n">install</span> <span class="n">brewsci</span><span class="o">/</span><span class="n">bio</span><span class="o">/</span><span class="n">muscle</span>
</pre></div>
</div>
<p>MUSCLE is easy to run, you just need to give it an input (<code class="docutils literal notranslate"><span class="pre">-in</span></code>) and output (<code class="docutils literal notranslate"><span class="pre">-out</span></code>) file. In our case, the input file is “Lesson_5_3_PAU5_BLAST_Hits.fasta”. The output file name could be anything you want:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">muscle</span> <span class="o">-</span><span class="ow">in</span> <span class="n">Lesson_5_3_PAU5_BLAST_Hits</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">out</span> <span class="n">Lesson_5_4_PAU5_MUSCLE</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
<p>The MUSCLE algorithm is an iterative method. By default it tries eight iterations of refinement. It should complete the alignment pretty fast with this small dataset.</p>
</div>
</div>
<div class="section" id="visualize-your-msa">
<h2>5.7. Visualize your MSA<a class="headerlink" href="#visualize-your-msa" title="Permalink to this headline">¶</a></h2>
<p>You could look at the results of your MUSCLE alignment in Terminal or a text editor, but it is going to be very hard to determine if the alignment is any good. While I generally use command line tools in my lab, sometimes you really need a GUI (graphical use interface).</p>
<p>There are many programs you can use to visualize MSAs, one option with an easy interface is <a class="reference external" href="https://alignmentviewer.org">ALIGNMENTVIEWER</a>. Once you load the data into the program you should see an alignment like this:</p>
<img src="Additional_Materials/Images/5_alignmentviewer.png">
<p>Scroll left to right to see the entire alignment. Some of the regions look pretty good, although they don’t look good for every sequence. Other regions look pretty bad. Assuming all of these sequences are actually Seripauperin proteins (in other words, they are all homologous), the poorly aligned regions represent mutations in some sequences that are difficult (perhaps impossible) to align with other sequences.</p>
<p>To get a trustworthy MSA, you should focus on the regions of each sequence that you can align with confidence. These poorly aligned regions might lead you to the wrong conclusion in downstream analysis. This is why <strong>cleaning up the data</strong> is usually a good idea before running a MSA.</p>
</div>
<div class="section" id="cleanup-option-1-only-use-the-aligned-regions-of-the-blast-hits">
<h2>5.8 Cleanup option 1: Only use the aligned regions of the BLAST hits<a class="headerlink" href="#cleanup-option-1-only-use-the-aligned-regions-of-the-blast-hits" title="Permalink to this headline">¶</a></h2>
<p>Let’s retry making our alignment. Instead of just extracting the sequence IDs from our BLAST analysis (the second column in file “Lesson_5_2_BLAST_Results.txt”), this time we are also going to extract the start and end coordinates (<code class="docutils literal notranslate"><span class="pre">sstart</span></code> and <code class="docutils literal notranslate"><span class="pre">send</span></code>) of the BLAST alignment. These are provided in the 9th and 10th columns in “Lesson_5_2_BLAST_Results.txt”.</p>
<p>Remember, if we want to extract coordinates with Samtools, we need the following syntax:</p>
<p>samtools faidx “sequence_ID”:start_coordinate-end_coordinate</p>
<p>We can get this format uaing <code class="docutils literal notranslate"><span class="pre">awk</span></code> but it takes a little thought. We can get awk to print the colon and dash by putting them in quotes (<code class="docutils literal notranslate"><span class="pre">&quot;:&quot;</span></code> and <code class="docutils literal notranslate"><span class="pre">&quot;-&quot;</span></code>). But what about the quotes themselves? Putting quotes in quotes (e.g. <code class="docutils literal notranslate"><span class="pre">&quot;&quot;&quot;</span></code>) will not work. Instead, we can invoke the octal code for double-quotes (<code class="docutils literal notranslate"><span class="pre">\042</span></code>)</p>
<p>Here’s the piece  of code to extract the BLAST hits and their matching coordinates with awk:</p>
<p>Extract the columns 2,9, and 19 from “Lesson_5_2_BLAST_Results.txt” with awk, adding the octal code for double quotes (<code class="docutils literal notranslate"><span class="pre">&quot;\042&quot;</span></code>):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">awk</span> <span class="s1">&#39;{print &quot;</span><span class="se">\042</span><span class="s1">&quot;$2&quot;</span><span class="se">\042</span><span class="s1">&quot;&quot;:&quot;$9&quot;-&quot;$10}&#39;</span> <span class="n">Lesson_5_2_BLAST_Results</span><span class="o">.</span><span class="n">txt</span> <span class="o">&gt;</span> <span class="n">Temp_BLAST_List</span>
</pre></div>
</div>
<p>Use the list “Temp_BLAST_List” to extract the sequence coordinates with Samtools:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">xargs</span> <span class="n">samtools</span> <span class="n">faidx</span> <span class="n">Lesson_4_Yeast_Proteome</span><span class="o">.</span><span class="n">fasta</span> <span class="o">&lt;</span> <span class="n">Temp_BLAST_List</span> <span class="o">&gt;</span> <span class="n">Lesson_5_5_PAU5_BLAST_Alignments</span><span class="o">.</span><span class="n">fasta</span>

<span class="n">rm</span> <span class="o">-</span><span class="n">i</span> <span class="n">Temp_BLAST_List</span>
</pre></div>
</div>
<p>Now run MUSCLE one more time:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">muscle</span> <span class="o">-</span><span class="ow">in</span> <span class="n">Lesson_5_5_PAU5_BLAST_Alignments</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">out</span> <span class="n">Lesson_5_6_MUSCLE_Alignments</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
<p>Load these new results into <a class="reference external" href="https://alignmentviewer.org">ALIGNMENTVIEWER</a> and see how it compares to the last alignment. It is definitely better, but there are some proteins that still align poorly. If you look at the sequence IDs you will notice that some have names such as “Cell_wall_protein” and “Temperature_shock-inducible_protein” instead of “Seripauperin”. This is a sign that some of the proteins from our BLAST search are not actually homologous, but instead are similar to the Seripauperin query we used by chance. Ideally, we want to get rid of those sequences too.</p>
</div>
<div class="section" id="cleanup-option-2-identify-and-isolate-conserved-domains">
<h2>5.9 Cleanup option 2: Identify and Isolate Conserved Domains<a class="headerlink" href="#cleanup-option-2-identify-and-isolate-conserved-domains" title="Permalink to this headline">¶</a></h2>
<div class="section" id="introducing-conserved-domains">
<h3>5.9.1. Introducing conserved domains<a class="headerlink" href="#introducing-conserved-domains" title="Permalink to this headline">¶</a></h3>
<p>Proteins play specific functions in biology. Proteins are often modular in their function, and some modules are more critical than others. Parts of the protein that are critical for specific tasks (e.g. binding to DNA, performing a catalytic reaction, interacting with another protein) often remain under strict selective pressure even as the rest of the protein evolves. The result is that homologous proteins often have highly conserved regions called <strong>conserved protein domains</strong>. Proteins that share a known conserved protein domain are almost certainly homologous.</p>
</div>
<div class="section" id="introducing-the-pfam-database">
<h3>5.9.2. Introducing the PFAM database<a class="headerlink" href="#introducing-the-pfam-database" title="Permalink to this headline">¶</a></h3>
<p><a class="reference external" href="https://pfam.xfam.org">The Pfam database</a> is a large collection of protein families, with information on known conserved protein domains. We can use our original Seripauperin query sequence to see if this protein has any conserved domains. I’m reprinting the query sequence below; use the <a class="reference external" href="https://pfam.xfam.org">Pfam Sequence Search</a> function to look for conserved domains:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;PAU5_Saccharomyces_arboricola_EJS43917.1
MVKLTSIAAGVAAIAAGASAATTTLAQSDEKVNLVELGVYVSDIRAHMAQYYLFQAAHPTETYPIEVAEA
VFNYGDFTTMLTGIAADQVTRMITGVPWYSTRLRPAISSALSKDGIYTIAN
</pre></div>
</div>
<p>The program will return one conserved domain, the “SRP1_TIP1 (Seripauperin and TIP1 family)” domain. When looking for Seripauperins in the <em>Saccharomyces cerevisiae</em> proteome, we would be wise to vet the BLAST results and only keep sequences that contain this domain.</p>
</div>
<div class="section" id="find-conserved-domains-with-the-pfam-web-server">
<h3>5.9.3. Find conserved domains with the PFAM web server<a class="headerlink" href="#find-conserved-domains-with-the-pfam-web-server" title="Permalink to this headline">¶</a></h3>
<p>You can identify conserved domains in multiple sequences using the <a class="reference external" href="https://pfam.xfam.org/search#tabview=tab1">Pfam web server</a>.</p>
<p>Load the <strong>original</strong> sequences from the BLAST search (Lesson_5_3_PAU5_BLAST_Hits.fasta) into PFAMscan. If you provide your email address you will receive the results in plain text.</p>
<p>To make life easier I have reproduced the results below and provide them as a text file (Lesson_5_PFAM_Hits.txt)</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="text-align:left head"><p>seq id</p></th>
<th class="text-align:left head"><p>alignment start</p></th>
<th class="text-align:left head"><p>alignment end</p></th>
<th class="text-align:left head"><p>envelope start</p></th>
<th class="text-align:left head"><p>envelope end</p></th>
<th class="text-align:left head"><p>hmm acc</p></th>
<th class="text-align:left head"><p>hmm name</p></th>
<th class="text-align:left head"><p>hmm start</p></th>
<th class="text-align:left head"><p>hmm end</p></th>
<th class="text-align:left head"><p>hmm length</p></th>
<th class="text-align:left head"><p>bit score</p></th>
<th class="text-align:left head"><p>Individual E-value</p></th>
<th class="text-align:left head"><p>Conditional E-value</p></th>
<th class="text-align:left head"><p>database significant</p></th>
<th class="text-align:left head"><p>outcompeted</p></th>
<th class="text-align:left head"><p>clan</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-align:left"><p>sp|Q08322|PAU20_YEAST_Seripauperin-20</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>114</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>115</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>94.38</p></td>
<td class="text-align:left"><p>3.5e-27</p></td>
<td class="text-align:left"><p>1.9e-31</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p>sp|P35994|PAU16_YEAST_Seripauperin-16</p></td>
<td class="text-align:left"><p>26</p></td>
<td class="text-align:left"><p>117</p></td>
<td class="text-align:left"><p>25</p></td>
<td class="text-align:left"><p>118</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>2</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>89.28</p></td>
<td class="text-align:left"><p>1.3e-25</p></td>
<td class="text-align:left"><p>7.5e-30</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p>sp|P0CE92|PAU8_YEAST_Seripauperin-8</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>114</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>115</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>95.40</p></td>
<td class="text-align:left"><p>1.7e-27</p></td>
<td class="text-align:left"><p>9.3e-32</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p>sp|P0CE93|PAU11_YEAST_Seripauperin-11</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>114</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>115</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>95.40</p></td>
<td class="text-align:left"><p>1.7e-27</p></td>
<td class="text-align:left"><p>9.3e-32</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p>sp|Q3E770|PAU9_YEAST_Seripauperin-9</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>114</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>115</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>95.40</p></td>
<td class="text-align:left"><p>1.7e-27</p></td>
<td class="text-align:left"><p>9.3e-32</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p>sp|P0CE91|PAU18_YEAST_Seripauperin-18</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>114</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>115</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>94.13</p></td>
<td class="text-align:left"><p>4.1e-27</p></td>
<td class="text-align:left"><p>2.3e-31</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p>sp|P0CE90|PAU6_YEAST_Seripauperin-6</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>114</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>115</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>94.13</p></td>
<td class="text-align:left"><p>4.1e-27</p></td>
<td class="text-align:left"><p>2.3e-31</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p>sp|P0CE89|PAU14_YEAST_Seripauperin-14</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>114</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>115</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>95.64</p></td>
<td class="text-align:left"><p>1.4e-27</p></td>
<td class="text-align:left"><p>7.8e-32</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p>sp|P0CE88|PAU1_YEAST_Seripauperin-1</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>114</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>115</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>95.64</p></td>
<td class="text-align:left"><p>1.4e-27</p></td>
<td class="text-align:left"><p>7.8e-32</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p>sp|P32612|PAU2_YEAST_Seripauperin-2</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>114</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>115</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>97.31</p></td>
<td class="text-align:left"><p>4.2e-28</p></td>
<td class="text-align:left"><p>2.4e-32</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p>sp|P53427|PAU4_YEAST_Seripauperin-4</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>114</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>115</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>94.86</p></td>
<td class="text-align:left"><p>2.5e-27</p></td>
<td class="text-align:left"><p>1.4e-31</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p>sp|P38725|PAU13_YEAST_Seripauperin-13</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>114</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>115</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>95.40</p></td>
<td class="text-align:left"><p>1.7e-27</p></td>
<td class="text-align:left"><p>9.3e-32</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p>sp|P0CE85|PAU19_YEAST_Seripauperin-19</p></td>
<td class="text-align:left"><p>26</p></td>
<td class="text-align:left"><p>117</p></td>
<td class="text-align:left"><p>25</p></td>
<td class="text-align:left"><p>118</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>2</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>89.17</p></td>
<td class="text-align:left"><p>1.5e-25</p></td>
<td class="text-align:left"><p>8.2e-30</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p>sp|P53343|PAU12_YEAST_Seripauperin-12</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>114</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>115</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>97.49</p></td>
<td class="text-align:left"><p>3.7e-28</p></td>
<td class="text-align:left"><p>2.1e-32</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p>sp|Q07987|PAU23_YEAST_Seripauperin-23</p></td>
<td class="text-align:left"><p>26</p></td>
<td class="text-align:left"><p>117</p></td>
<td class="text-align:left"><p>25</p></td>
<td class="text-align:left"><p>118</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>2</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>90.10</p></td>
<td class="text-align:left"><p>7.5e-26</p></td>
<td class="text-align:left"><p>4.2e-30</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p>sp|P0CE87|PAU22_YEAST_Seripauperin-22</p></td>
<td class="text-align:left"><p>66</p></td>
<td class="text-align:left"><p>157</p></td>
<td class="text-align:left"><p>65</p></td>
<td class="text-align:left"><p>158</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>2</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>88.00</p></td>
<td class="text-align:left"><p>3.4e-25</p></td>
<td class="text-align:left"><p>1.9e-29</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p>sp|P0CE86|PAU21_YEAST_Seripauperin-21</p></td>
<td class="text-align:left"><p>66</p></td>
<td class="text-align:left"><p>157</p></td>
<td class="text-align:left"><p>65</p></td>
<td class="text-align:left"><p>158</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>2</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>88.00</p></td>
<td class="text-align:left"><p>3.4e-25</p></td>
<td class="text-align:left"><p>1.9e-29</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p>sp|P40585|PAU15_YEAST_Seripauperin-15</p></td>
<td class="text-align:left"><p>26</p></td>
<td class="text-align:left"><p>117</p></td>
<td class="text-align:left"><p>25</p></td>
<td class="text-align:left"><p>118</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>2</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>89.24</p></td>
<td class="text-align:left"><p>1.4e-25</p></td>
<td class="text-align:left"><p>7.7e-30</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p>sp|Q03050|PAU10_YEAST_Seripauperin-10</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>113</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>115</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>97</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>94.74</p></td>
<td class="text-align:left"><p>2.7e-27</p></td>
<td class="text-align:left"><p>1.5e-31</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p>sp|P25610|PAU3_YEAST_Seripauperin-3</p></td>
<td class="text-align:left"><p>26</p></td>
<td class="text-align:left"><p>117</p></td>
<td class="text-align:left"><p>25</p></td>
<td class="text-align:left"><p>118</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>2</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>89.17</p></td>
<td class="text-align:left"><p>1.5e-25</p></td>
<td class="text-align:left"><p>8.2e-30</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p>sp|P38155|PAU24_YEAST_Seripauperin-24</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>114</p></td>
<td class="text-align:left"><p>22</p></td>
<td class="text-align:left"><p>115</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>96.29</p></td>
<td class="text-align:left"><p>8.8e-28</p></td>
<td class="text-align:left"><p>4.9e-32</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p>sp|P43575|PAU5_YEAST_Seripauperin-5</p></td>
<td class="text-align:left"><p>25</p></td>
<td class="text-align:left"><p>116</p></td>
<td class="text-align:left"><p>24</p></td>
<td class="text-align:left"><p>117</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>2</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>94.73</p></td>
<td class="text-align:left"><p>2.7e-27</p></td>
<td class="text-align:left"><p>1.5e-31</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p>sp|Q12370|PAU17_YEAST_Seripauperin-17</p></td>
<td class="text-align:left"><p>26</p></td>
<td class="text-align:left"><p>117</p></td>
<td class="text-align:left"><p>25</p></td>
<td class="text-align:left"><p>118</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>2</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>90.23</p></td>
<td class="text-align:left"><p>6.8e-26</p></td>
<td class="text-align:left"><p>3.8e-30</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p>sp|P47179|DAN4_YEAST_Cell_wall_protein_DAN4</p></td>
<td class="text-align:left"><p>25</p></td>
<td class="text-align:left"><p>116</p></td>
<td class="text-align:left"><p>24</p></td>
<td class="text-align:left"><p>117</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>2</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>80.61</p></td>
<td class="text-align:left"><p>6.8e-23</p></td>
<td class="text-align:left"><p>3.8e-27</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p>sp|P47178|DAN1_YEAST_Cell_wall_protein_DAN1</p></td>
<td class="text-align:left"><p>25</p></td>
<td class="text-align:left"><p>116</p></td>
<td class="text-align:left"><p>24</p></td>
<td class="text-align:left"><p>117</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>2</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>84.12</p></td>
<td class="text-align:left"><p>5.5e-24</p></td>
<td class="text-align:left"><p>3.1e-28</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p>sp|P39545|PAU7_YEAST_Seripauperin-7</p></td>
<td class="text-align:left"><p>25</p></td>
<td class="text-align:left"><p>55</p></td>
<td class="text-align:left"><p>23</p></td>
<td class="text-align:left"><p>55</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>2</p></td>
<td class="text-align:left"><p>32</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>25.30</p></td>
<td class="text-align:left"><p>1.2e-05</p></td>
<td class="text-align:left"><p>6.6e-10</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p>sp|P10863|TIR1_YEAST_Cold_shock-induced_protein_TIR1</p></td>
<td class="text-align:left"><p>13</p></td>
<td class="text-align:left"><p>113</p></td>
<td class="text-align:left"><p>13</p></td>
<td class="text-align:left"><p>115</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>97</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>126.98</p></td>
<td class="text-align:left"><p>2.4e-37</p></td>
<td class="text-align:left"><p>2.7e-41</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p>sp|P10863|TIR1_YEAST_Cold_shock-induced_protein_TIR1</p></td>
<td class="text-align:left"><p>211</p></td>
<td class="text-align:left"><p>227</p></td>
<td class="text-align:left"><p>210</p></td>
<td class="text-align:left"><p>227</p></td>
<td class="text-align:left"><p>PF00399.19</p></td>
<td class="text-align:left"><p>PIR</p></td>
<td class="text-align:left"><p>2</p></td>
<td class="text-align:left"><p>18</p></td>
<td class="text-align:left"><p>18</p></td>
<td class="text-align:left"><p>28.13</p></td>
<td class="text-align:left"><p>1.0e-06</p></td>
<td class="text-align:left"><p>1.1e-10</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p>sp|P27654|TIP1_YEAST_Temperature_shock-inducible_protein_1</p></td>
<td class="text-align:left"><p>17</p></td>
<td class="text-align:left"><p>110</p></td>
<td class="text-align:left"><p>14</p></td>
<td class="text-align:left"><p>113</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>4</p></td>
<td class="text-align:left"><p>96</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>114.99</p></td>
<td class="text-align:left"><p>1.3e-33</p></td>
<td class="text-align:left"><p>7.3e-38</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p>sp|P40552|TIR3_YEAST_Cell_wall_protein_TIR3</p></td>
<td class="text-align:left"><p>20</p></td>
<td class="text-align:left"><p>115</p></td>
<td class="text-align:left"><p>10</p></td>
<td class="text-align:left"><p>116</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>3</p></td>
<td class="text-align:left"><p>98</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>131.90</p></td>
<td class="text-align:left"><p>7.0e-39</p></td>
<td class="text-align:left"><p>3.9e-43</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p>sp|P33890|TIR2_YEAST_Cold_shock-induced_protein_TIR2</p></td>
<td class="text-align:left"><p>13</p></td>
<td class="text-align:left"><p>113</p></td>
<td class="text-align:left"><p>13</p></td>
<td class="text-align:left"><p>115</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>97</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>124.86</p></td>
<td class="text-align:left"><p>1.1e-36</p></td>
<td class="text-align:left"><p>1.2e-40</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p>sp|P33890|TIR2_YEAST_Cold_shock-induced_protein_TIR2</p></td>
<td class="text-align:left"><p>207</p></td>
<td class="text-align:left"><p>224</p></td>
<td class="text-align:left"><p>207</p></td>
<td class="text-align:left"><p>224</p></td>
<td class="text-align:left"><p>PF00399.19</p></td>
<td class="text-align:left"><p>PIR</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>18</p></td>
<td class="text-align:left"><p>18</p></td>
<td class="text-align:left"><p>28.45</p></td>
<td class="text-align:left"><p>8.1e-07</p></td>
<td class="text-align:left"><p>9.0e-11</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p>sp|Q12218|TIR4_YEAST_Cell_wall_protein_TIR4</p></td>
<td class="text-align:left"><p>13</p></td>
<td class="text-align:left"><p>113</p></td>
<td class="text-align:left"><p>13</p></td>
<td class="text-align:left"><p>116</p></td>
<td class="text-align:left"><p>PF00660.17</p></td>
<td class="text-align:left"><p>SRP1_TIP1</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>96</p></td>
<td class="text-align:left"><p>99</p></td>
<td class="text-align:left"><p>109.15</p></td>
<td class="text-align:left"><p>8.6e-32</p></td>
<td class="text-align:left"><p>4.8e-36</p></td>
<td class="text-align:left"><p>1</p></td>
<td class="text-align:left"><p>0</p></td>
<td class="text-align:left"><p></p></td>
</tr>
</tbody>
</table>
<p>Several of the sequences lack a “SRP1_TIP1” domain. We don’t want to keep those. We can use <code class="docutils literal notranslate"><span class="pre">grep</span></code> to keep the lines that have this domain. We can then use <code class="docutils literal notranslate"><span class="pre">awk</span></code> and <code class="docutils literal notranslate"><span class="pre">xargs</span> <span class="pre">samtools</span> <span class="pre">faidx</span></code> to extract these domains, based on the “alignment start” and “alignment end” coordinates (columns 2 and 3 in the table above):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">grep</span> <span class="s1">&#39;SRP1_TIP1&#39;</span> <span class="n">Lesson_5_PFAM_Hits</span><span class="o">.</span><span class="n">txt</span> <span class="o">&gt;</span> <span class="n">Temp1</span><span class="o">.</span><span class="n">txt</span>

<span class="n">awk</span> <span class="s1">&#39;{print &quot;</span><span class="se">\042</span><span class="s1">&quot;$1&quot;</span><span class="se">\042</span><span class="s1">&quot;&quot;:&quot;$2&quot;-&quot;$3}&#39;</span> <span class="n">Temp1</span><span class="o">.</span><span class="n">txt</span> <span class="o">&gt;</span> <span class="n">Temp2</span><span class="o">.</span><span class="n">txt</span>

<span class="n">xargs</span> <span class="n">samtools</span> <span class="n">faidx</span> <span class="n">Lesson_4_Yeast_Proteome</span><span class="o">.</span><span class="n">fasta</span> <span class="o">&lt;</span> <span class="n">Temp2</span><span class="o">.</span><span class="n">txt</span> <span class="o">&gt;</span> <span class="n">Lesson_5_7_PAU5_PFAM</span><span class="o">.</span><span class="n">fasta</span>

<span class="n">muscle</span> <span class="o">-</span><span class="ow">in</span> <span class="n">Lesson_5_7_PAU5_PFAM</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">out</span> <span class="n">Lesson_5_8_PAU5_MUSCLE_PFAM</span><span class="o">.</span><span class="n">fasta</span>

<span class="n">rm</span> <span class="o">-</span><span class="n">i</span> <span class="n">Temp</span><span class="o">*</span>
</pre></div>
</div>
<p>Load these new results into <a class="reference external" href="https://alignmentviewer.org">ALIGNMENTVIEWER</a> and see how they compare.</p>
</div>
</div>
<div class="section" id="upload-changes-to-your-github-repository">
<h2>5.10. Upload changes to your GitHub repository<a class="headerlink" href="#upload-changes-to-your-github-repository" title="Permalink to this headline">¶</a></h2>
<p>Don’t forget to upload the changes you made to your forked GitHub account:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="o">../</span>
<span class="n">git</span> <span class="n">add</span> <span class="o">--</span><span class="nb">all</span>
<span class="n">git</span> <span class="n">commit</span> <span class="o">-</span><span class="n">m</span> <span class="s1">&#39;performed samtools exercise&#39;</span>
<span class="n">git</span> <span class="n">push</span>
</pre></div>
</div>
<!-- #endregion -->
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
            



<div class='prev-next-bottom'>
    
    <div id="prev">
        <a class="left-prev" href="Lesson4.html" title="previous page">
            <i class="prevnext-label fas fa-angle-left"></i>
            <div class="prevnext-info">
                <p class="prevnext-label">previous</p>
                <p class="prevnext-title">Lesson 4: Finding Homologous Genes with BLAST</p>
            </div>
        </a>
    </div>

</div>
        
        </div>
    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By David A. Gold<br/>
        
            &copy; Copyright 2021.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="_static/js/index.1c5a1a01449ed65a7b51.js"></script>

  
  </body>
</html>